### PR TITLE
✨ 💥   Add shared train/load backends

### PR DESCRIPTION
**Pull request by joerunde**
This PR is the beginning of @gabe-l-hart and @tjohnson31415 's vision of the future of backends

This is a **breaking configuration change**

This:
- Adds `SharedLoadBackend` and `SharedTrainBackend` base classes and the `local` implementations of them
- Consolidates the `module_backends.priority` and `module_backends.config` configurations as a list of objects that include the backend type and its config
- Splits the backend priorities between training and loading, `module_backends.load_priority` and `module_backends.train_priority`
- Allows multiple backends of the same type to be registered under a different name and with different configurations
- Deletes the `module_backends.enabled` flag that was only used by the runtime. (It was buggy anyway, since the core loader would run backend configuration even if `module_backends.enabled` was False)

This does not yet implement a `caikit.train` method that can use an appropriate shared train backend similar to how `caikit.load` does.

As a followup design issue, the split between "shared" and the original "not shared" backends is pretty awkward. We'd like to explore if the original backends are really required or if they can be posed as shared backends as well. This would probably mean walking back the original "distributed module" pattern where multiple backend implementations are registered against the same guid as a base module

### CODE REVIEW

**Review by evaline-ju — COMMENTED** [review]
Due to the size of this PR, definitely just glossing over all the moved code in `module.py`

File: caikit/core/module_type.py
Comment (evaline-ju) [inline caikit/core/module_type.py:217]: module?

File: tests/core/test_model_manager.py
Comment (gabe-l-hart) [reply tests/core/test_model_manager.py:406]: I think the goal is to eventually remove all "special" handling for `LOCAL` and just have the default values in `config.yml` pre-populated with `LOCAL`. Could certainly argue either way, though

File: caikit/core/model_manager.py
Comment (gabe-l-hart) [inline caikit/core/model_manager.py:156]: I know we talked about reviving the `unique_id` content hash for the case where the same content is being loaded from multiple paths. Just noting that here for posterity

File: tests/core/test_model_manager.py
Comment (joerunde) [reply tests/core/test_model_manager.py:406]: right, needs more context.

The idea was to get `aconfig` to handle lists so that we could set `LOCAL` in the base config and allow users to merge more priorities to the list ahead of `LOCAL`, instead of handling it separately in code here

File: caikit/core/module_backends/base.py
Comment (gkumbhat) [inline caikit/core/module_backends/base.py:73]: Just thinking out the train use-case:
1. Would train data massaging / packaging also be responsibility of train backend impl class? Like in case of training in a torch cluster, if it requires data in torch data loaders, would the conversion from potentially stream in `caikit.runtime` to torch data loader be responsibility of impl of `SharedTrainBackend`?
2. Would we always be running `caikit.module` based training code in the backend? Or are we going to support non-caikit module training in backend as well?
3. Does error / exception handling responsibility of train function? Considering below `TODO` around async training, since `train` here can return future, would we need to expose another semantics for checking on the status, fetching error etc?
4.  How would the `.save` be handled and the case of automatic loading of the model, like the `auto_load_trained_model` case?

**Review by gkumbhat — COMMENTED** [review]
Thanks @gabe-l-hart , @joerunde . Left some questions

File: caikit/core/module.py
Comment (gabe-l-hart) [inline caikit/core/module.py:732]: @alex-jw-brooks I'm pretty sure you owned this originally. To the best of my knowledge, this function is an implementation detail that's only used in the default implementation of `evaluate_quality`. Is that right?

Comment (alex-jw-brooks) [reply caikit/core/module.py:732]: Hey @gabe-l-hart! As far as I'm aware, that is correct

File: caikit/core/module_backends/base.py
Comment (gabe-l-hart) [reply caikit/core/module_backends/base.py:73]: >Would train data massaging / packaging also be responsibility of train backend impl class? Like in case of training in a torch cluster, if it requires data in torch data loaders, would the conversion from potentially stream in caikit.runtime to torch data loader be responsibility of impl of SharedTrainBackend?

I have a hard time imagining a `SharedTrainBackend` that wraps running in a `torch` cluster since that would only work for module implementations that are backed by `torch`. I would expect such usages to be non-shared backends that require backend-specific implementations for a remote `torch` cluster. Realistically, the data conversions would probably be very similar across models, so the backend could expose any utility functions needed to make this code reusable.

In general, though, for a truly `Shared` train backend, yes, any data massaging would be the responsibility of the backend itself. For example, a shared train backend that delegates to a remote instance of `caikit.runtime` would need to serialize the python args into a train request `grpc` call.

>Would we always be running caikit.module based training code in the backend? Or are we going to support non-caikit module training in backend as well?

Good question. I think the baseline is that we would always have any code encapsulated as a `caikit` module. It could be totally fair game to have a generic module that allows for arbitrary training against a given platform, though. Again, something like this would probably not be a `Shared` backend, though, since the only generic interface that a `SharedTrainBackend` can rely on is `ModuleBase`.

>Does error / exception handling responsibility of train function? Considering below TODO around async training, since train here can return future, would we need to expose another semantics for checking on the status, fetching error etc?

Currently this does not return a future, but we absolutely will need to have careful error handling semantics down the line. Since the goal of a `SharedTrainBackend` is to encapsulate platform-specific concerns for a given backend platform, I think we will likely want to define a set of standard training error flows and make sure that all `SharedTrainBackend` implementations follow them.

>How would the .save be handled and the case of automatic loading of the model, like the auto_load_trained_model case?

This will be the crux of the `orchestration` design once we get there. We'll need to make sure that the train and load backends are able to interoperate cleanly to enable `auto_load_trained_model` semantics, even when the model is being trained and loaded on divergent backends. Certainly some design work to do here to figure out how to validate configurations and perform the model handoff from training to inference.

>From runtime perspective, can a shared backend in itself do distribution of training (internal distribution) and manage its own cluster?

By definition, a `SharedTrainBackend` is performing `external` distribution (the module is the atomic unit being distributed). Hypothetically a given shared backend _could_ also be used for `internal` distribution (distribution of parts of the workload inside a given module). For example, I could imagine a generic `ray` backend that holds the connection to a cluster and can perform distribution of modules as `remote` (`actor`, `job`?) primitives as a `SharedTrainBackend`, but also allows individual `ray`-based module implementations to grab the client and perform internal distribution.

As for "and manage its own cluster," this is an interesting question. My gut feeling is that or anything that will be running in a truly "remote" sense, the deployment topology should live outside the backend implementation to avoid mixing runtime and deploy-time concerns. That said, you could certainly imagine something like how [`caikit-tgis-backend` runs its subprocess](https://github.com/caikit/caikit-tgis-backend/blob/main/caikit_tgis_backend/tgis_backend.py#L82) for cluster-based backends that want to be able to operate in "build your own" mode.

File: tests/core/test_module.py
Comment (gabe-l-hart) [inline tests/core/test_module.py:414]: I think we need to add a new test to exercise the ability to disambiguate multiple instances of the same backend by name. I'll add that now

File: tests/core/helpers.py
Comment (gabe-l-hart) [inline tests/core/helpers.py:69]: Wouldn't the call to `set_load_backend` take care of this?

File: tests/core/test_model_manager.py
Comment (gabe-l-hart) [inline tests/core/test_model_manager.py:542]: I think this could be `model.get_load_backend.name == loader_name` and skip that extra `setattr`

**Review by gabe-l-hart — COMMENTED** [review]
@joerunde a couple of tiny NITs on the tests using `TestLoader`, but otherwise, a huge thanks for getting this fleshed out!

File: caikit/core/module_backends/base.py
Comment (gkumbhat) [reply caikit/core/module_backends/base.py:73]: Thanks Gabe.  

I think we would need to document above definitions as "parameters" for designing shared backends, so that not everything ends up being one-off shared backends and the difference between shared and non-shared backend is evident.

Comment (gabe-l-hart) [reply caikit/core/module_backends/base.py:73]: 100% agree. We probably need a pretty robust ADR around this. I can start writing one up today, but we need to sort out numbering at some point @joerunde

**Review by gabe-l-hart — APPROVED** [review]
From my view, this looks ready to go. Then again, I wrote a lot of it, so let's get signoff from one other maintainer before hitting merge!

File: tests/core/test_model_manager.py
Comment (joerunde) [reply tests/core/test_model_manager.py]: hot

**Comment by joerunde** [comment]
I can't sign off because I opened it 🙃

**Comment by gabe-l-hart** [comment]
@joerunde @evaline-ju I think verbal signoff is plenty! If you're good, let's merge it.

**Pull request merged**
